### PR TITLE
BUG: fix padding an empty array in reflect mode.

### PR DESCRIPTION
**Pull request by irushchyshyn**
This is a follow up PR for #9599, which rejected padding an empty array with nothing.
(Part of #9560)

### CODE REVIEW

File: numpy/lib/tests/test_arraypad.py [hunk 1]
```diff
@@ -1016,6 +1021,8 @@ def test_check_negative_pad_width(self):
     def test_check_empty_array(self):
         assert_raises(ValueError, pad, [], 4, mode='reflect')
         assert_raises(ValueError, pad, np.ndarray(0), 4, mode='reflect')
+        assert_raises(ValueError, pad, np.zeros((0, 3)), ((1,), (0,)),
+                      mode='reflect')
```
Comment (eric-wieser) [inline numpy/lib/tests/test_arraypad.py]: Can you add a test of the allowed case too?

File: numpy/lib/arraypad.py [hunk 2]
```diff
@@ -1406,10 +1406,14 @@ def pad(array, pad_width, mode, **kwargs):
             newmat = _append_min(newmat, pad_after, chunk_after, axis)
 
     elif mode == 'reflect':
-        if narray.size == 0:
-            raise ValueError("There aren't any elements to reflect in `array`")
-
         for axis, (pad_before, pad_after) in enumerate(pad_width):
+            if newmat.shape[axis] == 0:
+                # Axes with non-zero padding cannot be empty.
+                if ((pad_before > 0) or (pad_after > 0)):
```
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: I'd probably omit all the parens here. `if (...):` is just bad python that is trying to be C.

File: numpy/lib/tests/test_arraypad.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/lib/tests/test_arraypad.py]: `pad(np.zeros((0, 3)), ((0,), (1,)), mode='reflect')` would do, I think?

Comment (eric-wieser) [reply numpy/lib/tests/test_arraypad.py]: Might want to assert the new shape too.

Comment (irushchyshyn) [reply numpy/lib/tests/test_arraypad.py]: I do not think it should go into `TestValueError` test suite, so I have added this exact case to `TestReflect` .

File: numpy/lib/arraypad.py (hunk 2 above)
Comment (irushchyshyn) [reply numpy/lib/arraypad.py]: Correct, fixed.

File: numpy/lib/arraypad.py [hunk 3]
```diff
@@ -1406,10 +1406,14 @@ def pad(array, pad_width, mode, **kwargs):
...
 
     elif mode == 'reflect':
-        if narray.size == 0:
-            raise ValueError("There aren't any elements to reflect in `array`")
-
         for axis, (pad_before, pad_after) in enumerate(pad_width):
+            if newmat.shape[axis] == 0:
+                # Axes with non-zero padding cannot be empty.
+                if pad_before > 0 or pad_after > 0:
+                    raise ValueError("There aren't any elements to reflect in `array`")
```
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Perhaps this should be "in axis {} of `array`"?

**Review by eric-wieser — APPROVED** [review]
Looks good to me. One possible improvement, and waiting on you to squash the fixup

File: numpy/lib/arraypad.py (hunk 3 above)
Comment (irushchyshyn) [reply numpy/lib/arraypad.py]: Done.

File: numpy/lib/arraypad.py [hunk 4]
```diff
@@ -1406,10 +1406,15 @@ def pad(array, pad_width, mode, **kwargs):
             newmat = _append_min(newmat, pad_after, chunk_after, axis)
 
     elif mode == 'reflect':
-        if narray.size == 0:
-            raise ValueError("There aren't any elements to reflect in `array`")
-
         for axis, (pad_before, pad_after) in enumerate(pad_width):
+            if newmat.shape[axis] == 0:
```
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Why newmat and not narray as was used previously?

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: In fact, I don't see how the error here can be hit - newmat is the size after padding. Are you sure you're catching the right ValueError?

Comment (irushchyshyn) [reply numpy/lib/arraypad.py]: newmat is a copy of narray and I do the check before padding. The error will be hit if the axis it empty and the padding is non-zero. I might be wrong of course, so if you would insist on using `narray`, I do not mind.

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Since newmat.shape changes each iteration, I'd prefer narray. You're right though, both would work

Comment (irushchyshyn) [reply numpy/lib/arraypad.py]: Changed, thanks

**Comment by eric-wieser** [comment]
Thanks @irushchyshyn!

**Comment by irushchyshyn** [comment]
Thanks @eric-wieser for the review!

**Comment by sachin-r** [comment]
Dear numpy Community,

What is the ETA for the next release that will have this fix? We are using pip to install numpy, version returned is 1.13.1 which doesn't seems to have the fix.

Thanks

**Comment by charris** [comment]
1.13.2 will be out soon, but will not have this fix. I'm planning to branch 1.14.x fairly soon and would expect the first release candidate in a month or two. If you feel strongly that you need this fix in 1.13.2 let me know.

**Comment by sachin-r** [comment]
Thanks @charris for getting back. Our security team has blocked our product launch because of this  vulnerability (refer: https://www.cvedetails.com/cve/CVE-2017-12852/ ) , as numpy is used by almost by every other libraries we use, we can't find an alternative. We'd really appreciate a lot, if this fix is out soon

**Comment by charris** [comment]
OK, I've marked them for backport. If you would like to make a PR against `maintenance/1.13.x` for both of your fixes, that would be helpful.

**Comment by charris** [comment]
If you do make a PR, be sure to include the tests.

**Comment by sachin-r** [comment]
Thanks much @charris ! , Do I've to create a PR, I'm not familiar with process, Please guide me if I've to, thanks

**Comment by charris** [comment]
Ah, don't worry, in that case, I'll do it.

**Comment by sachin-r** [comment]
Thanks you so much!

**Comment by njsmith** [comment]
@sachin-r fyi the CVE is somewhat ridiculous. Unless you're actually calling `np.pad` with `mode="reflect"` and you let untrusted users control the size of the array being passed, then you're totally fine, and even if you do jump through the hoops to be vulnerable then all that would happen is a DoS. I guarantee that there are more serious bugs in most of your libraries that don't get assigned CVEs...

**Comment by charris** [comment]
@njsmith He can argue with the security team, or he can make a request to us. I suspect the latter is a *lot* easier.

**Comment by eric-wieser** [comment]
Note that numpy already has other problems that cause hangs anyway, involving recursive object arrays. For instance:

```
>>> x = np.array(None)
>>> x[()] = x
>>> x == x
```

That is far worse than the issue described here, as the loop is in C land, and can't even be interrupted by Ctrl C. I also suspect it causes a stack overflow.

**Comment by sachin-r** [comment]
@njsmith thanks for the input, in our code we can make sure that np.pad is not called with the aforementioned mode param but the Security team here seems to worry about that we use tensorflow, keras padding & other 3rd party libraries that may or may not be using np.pad in that mode. The security team came  across this CVE & worries about the DoS stated in the CVE. As we are large enterprise,  they are very conservative about security

**Comment by sachin-r** [comment]
@charris please share the process to create a request, if it would help this fix to be out in 1.13.2

**Comment by charris** [comment]
@sachin-r It's aready in 1.3.x and will be in the 1.3.2 release, which will come out in a couple of days.

@eric-wieser No doubt other things that need fixing, but having a CVE out there makes it something we need to deal with because it makes people nervous and looks bad if we don't. Corporations can be picky about such things ...

**Comment by charris** [comment]
@sachin-r For basic instruction on making pull requests, see https://docs.scipy.org/doc/numpy-1.13.0/dev/gitwash/git_development.html .

**Comment by sachin-r** [comment]
Thanks for understanding @charris , sorry to be pedantic but earlier in this thread it say this is removed from 1.13.2 release milestone. Just want to make sure if this is not something to worry about?
image

**Comment by charris** [comment]
Yep, this PR was removed because another against 1.13.x was added. See #9742.

**Comment by charris** [comment]
You will notice a reference to the new PR up above before the tags were removed.

**Comment by sachin-r** [comment]
Thanks for clarifying, Have a nice weekend!
